### PR TITLE
fix(terraform): do not evaluate OCI policy statements

### PR DESCRIPTION
**Pull request by behnazh-w**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally we encourage adding a scope to the prefix, which indicates the targeted framework, otherwise 'general' will be assumed.
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

Checkov currently evaluates the `statements` attributes of a Terraform resource, which contain IAM policies. However, the syntax of the policies differ for different providers, which results in unexpected evaluations. Oracle Cloud Infrastructure uses a [special syntax](https://docs.oracle.com/en-us/iaas/Content/Identity/Concepts/policysyntax.htm) for the policy statements, but Checkov seems to try to evaluate it as a JSON object. 

Here is a solution which prevents evaluating the `statements` attributes when the provider is set to `OCI` through an environment variable.

Closes https://github.com/bridgecrewio/checkov/issues/3412

### CODE REVIEW

**Comment by behnazh-w** [comment]
@gruebel Is there anything I can do on my end to help merging this PR?

**Comment by gruebel** [comment]
hey @behnazh-w I adjusted your code a bit. Please have a look and you can remove your change in the GitHub workflow file 🙂

File: checkov/terraform/graph_builder/variable_rendering/renderer.py [hunk 1]
```diff
@@ -25,8 +27,7 @@
 if TYPE_CHECKING:
     from checkov.terraform.graph_builder.local_graph import TerraformLocalGraph
 
-ATTRIBUTES_NO_EVAL = ["template_body", "template"]
-VAR_TYPE_DEFAULT_VALUES = {
+VAR_TYPE_DEFAULT_VALUES: dict[str, list[Any] | dict[str, Any]] = {
```
Comment (behnazh-w) [inline checkov/terraform/graph_builder/variable_rendering/renderer.py:30]: @gruebel Because these changes seem to be type and linting fixes that are not related to my changes, do you mind if I add them in a separate commit to have a clean history? That would mean that the two commits shouldn't be squashed though in this PR.

File: checkov/terraform/graph_builder/variable_rendering/renderer.py [hunk 2]
```diff
@@ -1,3 +1,5 @@
+from __future__ import annotations
```
Comment (behnazh-w) [inline checkov/terraform/graph_builder/variable_rendering/renderer.py:1]: Just out of curiosity, what's the purpose of importing `annotations` here? Is it for performance reasons as described in [PEP 563](https://peps.python.org/pep-0563/)?

File: tests/terraform/graph/graph_builder/test_oci_policy.py [hunk 3]
```diff
@@ -1,43 +1,38 @@
...
+from pathlib import Path
+from unittest import mock
 
-from checkov.common.graph.db_connectors.networkx.networkx_db_connector import (
-    NetworkxConnector,
-)
+from checkov.common.graph.db_connectors.networkx.networkx_db_connector import NetworkxConnector
 from checkov.terraform.graph_manager import TerraformGraphManager
 
+@mock.patch.dict(os.environ, {"CKV_TERRAFORM_PROVIDER": "OCI"})
```
Comment (behnazh-w) [inline tests/terraform/graph/graph_builder/test_oci_policy.py:10]: I also considered mocking the `env` variable but because `attributes_no_eval` is extended at module load time, the test wouldn't have exercised my change exactly. So I decided to set the `env` variable before running `pytest` in the GHA workflow to make sure it doesn't have any side effects. If you think mocking is enough for testing, I'm fine with it.

File: tests/terraform/graph/graph_builder/test_oci_policy.py [hunk 4]
```diff
@@ -1,43 +1,38 @@
...
+def test_oci_policy_statements_with_provider_env_var():
```
Comment (behnazh-w) [inline tests/terraform/graph/graph_builder/test_oci_policy.py:11]: I followed the other tests and created a class to be consistent. But if a method is enough, I'm happy with that.

File: tests/terraform/graph/graph_builder/test_oci_policy.py [hunk 5]
```diff
@@ -1,43 +1,38 @@
...
-                str(
-                    local_graph.vertices[0].config["oci_identity_policy"]["example"][
-                        "statements"
-                    ]
-                ),
-                "[[\"allow group group-admin-001 to use groups in tenancy where target.group.name != 'Administrators'\"]]",
-            )
+    # then
+    statements = local_graph.vertices[0].config["oci_identity_policy"]["example"]["statements"][0]
+    assert statements == ["True"]
```
Comment (behnazh-w) [inline tests/terraform/graph/graph_builder/test_oci_policy.py:38]: I can add a comment to clarify that `"True"` is a wrong value due to special syntax in OCI policy.

**Comment by behnazh-w** [comment]
@gruebel Thanks a lot for making the effort. I have left some comments and questions.

**Comment by behnazh-w** [comment]
@gruebel thanks again for reviewing this PR. I separated the changes to two commits: first one is for fixing the linting issues and the second one fixes the policy evaluation. 

It's ready to be merged. Just a reminder not to squash the two commits to avoid mixing the linting fixes with the policy fix.

File: checkov/terraform/graph_builder/variable_rendering/renderer.py (hunk 2 above)
Comment (gruebel) [reply checkov/terraform/graph_builder/variable_rendering/renderer.py:1]: mostly for better and easier type hinting 🙂 you can use built-in types instead of the typing variants, like `list` instead of `List`.

File: tests/terraform/graph/graph_builder/test_oci_policy.py (hunk 4 above)
Comment (gruebel) [reply tests/terraform/graph/graph_builder/test_oci_policy.py:11]: yeah it is kind of mixed. we started with `untitest` style and still do them, but as a test runner we use `pytest` and then you actually can just use functions, which I prefer 🙂

File: tests/terraform/graph/graph_builder/test_oci_policy.py (hunk 3 above)
Comment (gruebel) [reply tests/terraform/graph/graph_builder/test_oci_policy.py:10]: yeah, it only works because I changed the global variable to be an instance variable, otherwise this approach wouldn't work, because they are only loaded once.

File: checkov/terraform/graph_builder/variable_rendering/renderer.py [hunk 6]
```diff
@@ -40,6 +41,11 @@ class TerraformVariableRenderer(VariableRenderer):
     def __init__(self, local_graph: "TerraformLocalGraph") -> None:
         super().__init__(local_graph)
 
+        self.attributes_no_eval = {"template_body", "template"}
+        if os.environ.get("CKV_TERRAFORM_PROVIDER") == "OCI":
```
[the remark this answers is not in the document]
Comment (nimrodkor) [reply checkov/terraform/graph_builder/variable_rendering/renderer.py:45]: Suggestion - have attributes_no_eval be a method, that gets the vertex index, and checks if the vertex at this index is an OCI terraform resource - starts with `oci_`. If so - does not render statements

[the remark this answers is not in the document]
Comment (behnazh-w) [reply checkov/terraform/graph_builder/variable_rendering/renderer.py]: > Suggestion - have attributes_no_eval be a method, that gets the vertex index, and checks if the vertex at this index is an OCI terraform resource - starts with `oci_`. If so - does not render statements

@nimrodkor Thanks for the suggestion.  Please see commit 1016cdacf0f545210b24c727c1986b609df8ab01.

**Review by nimrodkor — APPROVED** [review]

File: checkov/terraform/graph_builder/variable_rendering/renderer.py [hunk 7]
```diff
@@ -40,6 +41,26 @@ class TerraformVariableRenderer(VariableRenderer):
...
+        :param attribute: the attribute to check
+        :param vertex_index: the index of the current vertex
+        :return bool: True if the attribute should not be evaluated and False otherwise
+        """
+        if attribute in {"template_body", "template"}:
+            return True
+
+        # OCI policy statements have a special syntax and should not be evaluated.
+        # Check if the vertex at this index is an OCI terraform resource.
+        vertex_attributes = self.local_graph.get_vertex_attributes_by_index(vertex_index)
```
Comment (rotemavni) [inline checkov/terraform/graph_builder/variable_rendering/renderer.py:57]: @behnazh-w How do you feel about fetching the attributes only if `attribute == "statements"`? I would be more efficient (this way we don't access the graph for attributes that should not be skipped)

**Review by rotemavni — COMMENTED** [review]
Thank you for you contribution! Had just one comment.

File: checkov/terraform/graph_builder/variable_rendering/renderer.py (hunk 7 above)
Comment (behnazh-w) [reply checkov/terraform/graph_builder/variable_rendering/renderer.py]: > @behnazh-w How do you feel about fetching the attributes only if `attribute == "statements"`? I would be more efficient (this way we don't access the graph for attributes that should not be skipped)

Thanks @rotemavni for the feedback. Please see commit eaeceb10c88a27b04050676d5f2144bacfcfcf5c.

**Review by rotemavni — APPROVED** [review]

File: checkov/terraform/graph_builder/variable_rendering/renderer.py (hunk 7 above)
Comment (rotemavni) [reply checkov/terraform/graph_builder/variable_rendering/renderer.py]: Thank you @behnazh-w, I approved the PR
